src/plugins/profiling/common/functions.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  compareFrameGroup,
+  defaultGroupBy,
+  Executable,
+  FileID,
+  FrameGroup,
+  FrameGroupID,
+  groupStackFrameMetadataByStackTrace,
+  hashFrameGroup,
+  StackFrame,
+  StackFrameID,
+  StackFrameMetadata,
+  StackTrace,
+  StackTraceID,
+} from './profiling';
+
+type TopNFunctionAndFrameGroup = {
+  Frame: StackFrameMetadata;
+  FrameGroup: FrameGroup;
+  CountExclusive: number;
+  CountInclusive: number;
+};
+
+type TopNFunction = Pick<TopNFunctionAndFrameGroup, 'Frame' | 'CountExclusive' | 'CountInclusive'>;
+
+type TopNFunctions = {
+  TotalCount: number;
+  TopN: TopNFunction[];
+};
+
+export function createTopNFunctions(
+  events: Map<StackTraceID, number>,
+  stackTraces: Map<StackTraceID, StackTrace>,
+  stackFrames: Map<StackFrameID, StackFrame>,
+  executables: Map<FileID, Executable>,
+  startIndex: number,
+  endIndex: number
+): TopNFunctions {
+  const metadata = groupStackFrameMetadataByStackTrace(stackTraces, stackFrames, executables);
+
+  // The `count` associated with a frame provides the total number of
+  // traces in which that node has appeared at least once. However, a
+  // frame may appear multiple times in a trace, and thus to avoid
+  // counting it multiple times we need to record the frames seen so
+  // far in each trace.
+  let totalCount = 0;
+  const topNFunctions = new Map<FrameGroupID, TopNFunctionAndFrameGroup>();
+
+  // Collect metadata and inclusive + exclusive counts for each distinct frame.
+  for (const [traceHash, count] of events) {
+    const uniqueFrameGroupsPerEvent = new Set<FrameGroupID>();
+
+    totalCount += count;
+
+    const frames = metadata.get(traceHash)!;
+    for (let i = 0; i < frames.length; i++) {
```

Comment:
In this case we need a special handling for the last element further down.
And in general I personally prefer explicit for loops in Typescript and forEach in JS.
What is the general consent in Elastic about this ? Is there a style guide or recommendations ?

Reply:
I think `forEach` is more common. There's an explicit lint rule against `for(... in ...)` too.

Reply:
Since I work most of the time with Go, C/C++, Java, the common denominator is a for loop.
Indeed, for me it is much easier to read an explicit for loop than a forEach construction.
And therefore I like Google's recommendations for idiomatic Typescript: https://google.github.io/styleguide/tsguide.html#iterating-containers
😃